File: examples/js/websocket-poloniex-orderbook.js
'use strict';

const asTable = require('as-table'),
	log = require('ololog').noLocate,
	ansi = require('ansicolor').nice,
	ccxt = require('../../ccxt.js'),
	cex = require('../../js/cex');



let sleep = ms => new Promise(resolve => setTimeout(resolve, ms));

let exchange;
async function fetchOrderBook(id, apiKey, secret, limit, symbol, params) {
	exchange = new ccxt[id]({
		apiKey: apiKey,
		secret: secret,
		enableRateLimit: true,
		verbose: false,
	});
	exchange.on('err', (err, conxid) => {
		try {
			console.log(err);
			exchange.websocketClose(conxid);
			return
		} catch (ex) {
			console.log(ex);
		}
	});
	exchange.on('ob', (market, ob) => {
		console.log(' ');
		console.log('ob updated: ', market, ob);
	});
	await exchange.loadMarkets();

	console.log('subscribe: ' + symbol);
	await exchange.websocketSubscribe('ob', symbol, params);
	console.log('subscribed: ' + symbol);

	await exchange.websocketFetchOrderBook(symbol, limit);
	
	await sleep(15 * 1000);

	console.log('unsubscribe: ' + symbol);
	await exchange.websocketUnsubscribe('ob', symbol, params);
	console.log('unsubscribed: ' + symbol);
}

(async function main() {
	try {
		const id = 'poloniex';
		const apiKey = 'do not care';
		const secret = 'do not care';
		const limit = 2;
		const obDeltaCacheSizeMax = 10
		const symbols = 'ETH/BTC';

		const ob = await fetchOrderBook(id, apiKey, secret, limit, symbols, {
			// contract_type: 'next_week',
			'limit': limit,
			'obDeltaCacheSizeMax': obDeltaCacheSizeMax,
		});
	} catch (ex) {
		console.log('MAIN() EXCEPTION: ')
		log('Error:'.red, ex);
		log(ex.stack);
		exchange.websocketClose();
	}
})();

File: examples/py/websocket-poloniex-orderbook.py
import os
import sys
import pprint
import traceback

pp = pprint.PrettyPrinter(depth=6)

root = os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
sys.path.append(root + '/python')
# import ccxt  # noqa: E402
import ccxt.async_support as ccxt  # noqa: E402
import asyncio  # noqa: E402

loop = asyncio.get_event_loop()
# import txaio
# txaio.start_logging(level='debug')


async def main():

    exchange_id = 'poloniex'
    apiKey = 'do not care'
    secret = 'do not care'
    limit = 2
    symbol = 'ETH/BTC'

    exchange = getattr(ccxt, exchange_id)({
        "apiKey": apiKey,
        "secret": secret,
        "enableRateLimit": True,
        'verbose': False,
        'timeout': 5 * 1000,
        # 'wsproxy': 'http://185.93.3.123:8080/',
    })

    @exchange.on('err')
    def websocket_error(err, conxid):  # pylint: disable=W0612
        print(type(err).__name__ + ":" + str(err))
        traceback.print_tb(err.__traceback__)
        traceback.print_stack()
        loop.stop()

    @exchange.on('ob')
    def websocket_ob(symbol, ob):  # pylint: disable=W0612
        print("ob updated: " + symbol)
        sys.stdout.flush()
        pp.pprint(ob)

    sys.stdout.flush()

    print("subscribe: " + symbol)
    sys.stdout.flush()
    await exchange.websocket_subscribe('ob', symbol, {'limit': limit})
    print("subscribed: " + symbol)
    sys.stdout.flush()
    ob = await exchange.websocket_fetch_order_book(symbol, limit)  # noqa: F841 pylint: disable=W0612
    print("ob fetched: " + symbol)
    # print(ob)
    sys.stdout.flush()
    await asyncio.sleep(15)

    print("unsubscribe: " + symbol)
    sys.stdout.flush()
    await exchange.websocket_unsubscribe('ob', symbol)
    print("unsubscribed: " + symbol)
    sys.stdout.flush()

    await exchange.close()

print("before start")
loop.run_until_complete(main())
# loop.run_forever()
# loop.stop()
# loop.close()
print("after complete")
